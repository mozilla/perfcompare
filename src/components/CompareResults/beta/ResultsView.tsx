import { useEffect } from 'react';

import type { Theme } from '@mui/material';
import Grid from '@mui/material/Grid';
import { useLocation } from 'react-router-dom';
import { style } from 'typestyle';

import useFetchCompareResults from '../../../hooks/useFetchCompareResults';
import useHandleChangeSearch from '../../../hooks/useHandleChangeSearch';
import { SearchContainerStyles } from '../../../styles';
import { background } from '../../../styles';
import { Repository } from '../../../types/state';
import CompareWithBase from '../../Search/CompareWithBase';
import PerfCompareHeader from '../../Shared/PerfCompareHeader';
import ResultsMain from './ResultsMain';

interface ResultsViewProps {
  protocolTheme: Theme;
  toggleColorMode: () => void;
}

function ResultsView(props: ResultsViewProps) {
  const { protocolTheme, toggleColorMode } = props;
  const themeMode = protocolTheme.palette.mode;
  const view = 'compare-results';
  const styles = {
    container: style({
      backgroundColor: background(themeMode),
    }),
  };

  const sectionStyles = SearchContainerStyles(themeMode, view);

  const location = useLocation();
  const { dispatchFetchCompareResults } = useFetchCompareResults();
  const { searchByRevisionOrEmail } = useHandleChangeSearch();

  useEffect(() => {
    const searchParams = new URLSearchParams(location.search);
    const repos = searchParams.get('repos')?.split(',');
    const revs = searchParams.get('revs')?.split(',');
    void dispatchFetchCompareResults(repos as Repository['name'][], revs);

    if (revs && repos) {
      revs.forEach((rev, index) => {
        void searchByRevisionOrEmail(
          repos[index] as Repository['name'],
          rev,
          'base',
        );
        void searchByRevisionOrEmail(
          repos[index] as Repository['name'],
          rev,
          'new',
        );
      });
    }
  }, []);

  return (
    <div
      className={styles.container}
      data-testid='beta-version-compare-results'
    >
      <PerfCompareHeader
        themeMode={themeMode}
        toggleColorMode={toggleColorMode}
        view='compare-results'
      />
      <section className={sectionStyles.container}>
        <CompareWithBase mode={themeMode} view={view} />
      </section>
      <Grid container alignItems='center' justifyContent='center'>
        <Grid item xs={12}>
          <ResultsMain themeMode={themeMode} />
        </Grid>
      </Grid>
    </div>
  );
}

export default ResultsView;
