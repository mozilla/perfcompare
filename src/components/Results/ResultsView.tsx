import React from 'react';

import Container from '@mui/material/Container';
import Grid from '@mui/material/Grid';

import PerfCompareHeader from '../Shared/PerfCompareHeader';
import SelectedRevisionsTable from '../Shared/SelectedRevisionsTable';

function ResultsView() {
  return (
    <Container maxWidth="lg">
      <PerfCompareHeader />
      <Grid container>
        <Grid item xs={12}>
          <SelectedRevisionsTable />
        </Grid>
      </Grid>
    </Container>
  );
}

export default ResultsView;
