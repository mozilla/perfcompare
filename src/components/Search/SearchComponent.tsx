import { useState } from 'react';

import InfoIcon from '@mui/icons-material/InfoOutlined';
import Grid from '@mui/material/Grid';
import InputLabel from '@mui/material/InputLabel';
import Tooltip from '@mui/material/Tooltip';
import { cssRule } from 'typestyle';

import { compareView } from '../../common/constants';
import { useAppSelector } from '../../hooks/app';
import {
  Spacing,
  DropDownMenuRaw,
  DropDownItemRaw,
  //SearchStyles can be found in CompareCards.ts
  SearchStyles,
} from '../../styles';
import type { Changeset, InputType, Repository } from '../../types/state';
import EditButton from './EditButton';
import SaveCancelButtons from './SaveCancelButtons';
import SearchDropdown from './SearchDropdown';
import SearchInputAndResults from './SearchInputAndResults';
import SelectedRevisions from './SelectedRevisions';

interface SearchProps {
  hasNonEditableState: boolean;
  isWarning: boolean;
  isBaseComp: boolean;
  displayedRevisions: Changeset[];
  onSave: () => void;
  onCancel: () => void;
  onEdit: () => void;
  onSearchResultsToggle: (item: Changeset) => void;
  onRemoveRevision: (item: Changeset) => void;
  selectLabel: string;
  tooltip: string;
  inputPlaceholder: string;
  repository: Repository['name'];
  labelIdInfo: string;
  onRepositoryChange: (repo: Repository['name']) => unknown;
}

function SearchComponent({
  hasNonEditableState,
  isBaseComp,
  displayedRevisions,
  onCancel,
  onSave,
  onEdit,
  onSearchResultsToggle,
  onRemoveRevision,
  selectLabel,
  tooltip,
  inputPlaceholder,
  isWarning,
  repository,
  labelIdInfo,
  onRepositoryChange,
}: SearchProps) {
  const mode = useAppSelector((state) => state.theme.mode);
  const styles = SearchStyles(mode);
  const searchType: InputType = isBaseComp ? 'base' : 'new';

  /* These overriding rules update the theme mode by accessing the otherwise inaccessible MUI tooltip styles */
  cssRule('.MuiPopover-root', {
    $nest: {
      '.MuiPaper-root': {
        flexDirection: 'column',
        ...(mode === 'light' ? DropDownMenuRaw.Light : DropDownMenuRaw.Dark),
        $nest: {
          '.MuiList-root': {
            padding: `${Spacing.Small}px ${Spacing.xSmall}px`,
            $nest: {
              '.MuiMenuItem-root': {
                ...(mode === 'light'
                  ? DropDownItemRaw.Light
                  : DropDownItemRaw.Dark),
              },
            },
          },
        },
      },
    },
  });

  const [formIsDisplayed, setFormIsDisplayed] = useState(!hasNonEditableState);

  return (
    <Grid className={styles.component}>
      <Grid
        item
        xs={2}
        className={`${isBaseComp ? 'base' : 'new'}-search-dropdown ${
          styles.dropDown
        } label-edit-wrapper`}
      >
        <InputLabel id={labelIdInfo} className='dropdown-select-label'>
          {selectLabel}
          <Tooltip placement='top' title={tooltip}>
            <InfoIcon fontSize='small' className='dropdown-info-icon' />
          </Tooltip>
        </InputLabel>
        {/**** Edit Button ****/}
        {hasNonEditableState && !formIsDisplayed && (
          <EditButton
            isBase={isBaseComp}
            onEditAction={() => {
              onEdit();
              setFormIsDisplayed(true);
            }}
          />
        )}
      </Grid>
      {/**** Search - DropDown Section ****/}
      <Grid
        container
        alignItems='flex-start'
        id={`${searchType}-search-container`}
        className={`${styles.container} ${
          formIsDisplayed ? 'show-container' : 'hide-container'
        } `}
      >
        <Grid
          item
          xs={2}
          id={`${searchType}_search-dropdown`}
          className={`${searchType}-search-dropdown ${styles.dropDown} ${
            hasNonEditableState ? 'small' : ''
          } ${hasNonEditableState ? compareView : ''}-base-dropdown`}
        >
          <SearchDropdown
            compact={hasNonEditableState}
            selectLabel={selectLabel}
            searchType={searchType}
            repository={repository}
            labelIdInfo={labelIdInfo}
            onChange={onRepositoryChange}
          />
        </Grid>
        <Grid
          item
          xs={7}
          id={`${searchType}_search-input`}
          className={`${searchType}-search-input  ${styles.baseSearchInput} ${
            hasNonEditableState ? 'big' : ''
          } `}
        >
          <SearchInputAndResults
            compact={hasNonEditableState}
            inputPlaceholder={inputPlaceholder}
            displayedRevisions={displayedRevisions}
            searchType={searchType}
            repository={repository}
            onSearchResultsToggle={onSearchResultsToggle}
          />
        </Grid>
        {/****** Cancel Save Buttons ******/}
        {hasNonEditableState && formIsDisplayed && (
          <SaveCancelButtons
            searchType={searchType}
            onSave={() => {
              onSave();
              setFormIsDisplayed(false);
            }}
            onCancel={() => {
              onCancel();
              setFormIsDisplayed(false);
            }}
          />
        )}
      </Grid>
      {/***** Selected Revisions Section *****/}
      {displayedRevisions && (
        <Grid
          className='d-flex'
          data-testid={`${searchType}-selected-revision`}
        >
          <SelectedRevisions
            isBase={isBaseComp}
            canRemoveRevision={!hasNonEditableState || formIsDisplayed}
            isWarning={isWarning}
            displayedRevisions={displayedRevisions}
            onRemoveRevision={onRemoveRevision}
          />
        </Grid>
      )}
    </Grid>
  );
}

export default SearchComponent;
