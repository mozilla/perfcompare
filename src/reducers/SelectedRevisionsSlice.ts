import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { RevisionsList, SelectedRevisionsState } from '../types/state';

const initialState: SelectedRevisionsState = {
  revisions: [],
  baseCommittedRevisions: [],
  newCommittedRevisions: [],
};

const selectedRevisions = createSlice({
  name: 'selectedRevisions',
  initialState,
  reducers: {
    //when the user presses the "Compare"
    setSelectedRevisions(
      state,
      action: PayloadAction<{
        selectedRevisions: RevisionsList[];
      }>,
    ) {
      state.revisions = action.payload.selectedRevisions;
      state.baseCommittedRevisions = [action.payload.selectedRevisions[0]];
      //returns array without first element
      state.newCommittedRevisions = action.payload.selectedRevisions.slice(1);
    },
  },
});

export const { setSelectedRevisions } = selectedRevisions.actions;
export default selectedRevisions.reducer;
