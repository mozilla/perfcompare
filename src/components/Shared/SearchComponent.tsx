import {
  Dispatch,
  SetStateAction,
  useEffect,
  useState,
  createRef,
} from 'react';

import Grid from '@mui/material/Grid';
import useMediaQuery from '@mui/material/useMediaQuery';

import { RootState } from '../../common/store';
import { useAppDispatch } from '../../hooks/app';
import { useAppSelector } from '../../hooks/app';
import { clearCheckedRevisions } from '../../reducers/CheckedRevisions';
import { SearchStyles } from '../../styles';
import type { RevisionsList, InputType } from '../../types/state';
import SearchDropdown from '../Search/SearchDropdown';
import SearchInput from '../Search/SearchInput';
import SearchResultsList from '../Search/SearchResultsList';

interface SearchProps {
  mode: 'light' | 'dark';
  view: 'compare-results' | 'search';
  setPopoverIsOpen?: Dispatch<SetStateAction<boolean>>;
  prevRevision?: RevisionsList;
  selectLabel: string;
  tooltip: string;
  inputPlaceholder: string;
  searchType: InputType;
}

function SearchComponent({
  mode,
  view,
  selectLabel,
  tooltip,
  inputPlaceholder,
  searchType,
}: SearchProps) {
  const styles = SearchStyles(mode);
  const searchState = useAppSelector(
    (state: RootState) => state.search[searchType],
  );
  const { searchResults } = searchState;
  const [focused, setFocused] = useState(false);
  const dispatch = useAppDispatch();
  const matchesQuery = useMediaQuery('(max-width:768px)');
  const containerRef = createRef<HTMLDivElement>();

  const handleFocus = (e: MouseEvent) => {
    if (
      (e.target as HTMLElement).matches(`#${searchType}-search-container, 
      #${searchType}-search-container *`) &&
      // do not open search results when dropdown or cancel button is clicked
      !(e.target as HTMLElement).matches(
        `#${searchType}_search-dropdown,
        #${searchType}_search-dropdown *,
        #cancel-edit-${searchType}-button, 
        #cancel-edit-${searchType}-button *`,
      )
    ) {
      setFocused(true);
      return;
    } else {
      setFocused(false);
    }
  };

  const handleEscKeypress = (e: KeyboardEvent) => {
    if (e.key === 'Escape') {
      setFocused(false);
    }
  };

  useEffect(() => {
    if (focused == false) {
      dispatch(clearCheckedRevisions());
    }
  }, [focused]);

  useEffect(() => {
    document.addEventListener('mousedown', handleFocus);
    return () => {
      document.removeEventListener('mousedown', handleFocus);
    };
  }, []);

  useEffect(() => {
    document.addEventListener('keydown', handleEscKeypress);
    return () => {
      document.removeEventListener('keydown', handleEscKeypress);
    };
  });

  return (
    <Grid
      container
      alignItems='flex-start'
      id={`${searchType}-search-container`}
      className={styles.container}
      ref={containerRef}
    >
      <Grid
        item
        xs={2}
        id={`${searchType}_search-dropdown`}
        className={`${searchType}-search-dropdown ${styles.dropDown} ${styles.baseSearchDropdown}`}
      >
        <SearchDropdown
          view={view}
          selectLabel={selectLabel}
          tooltipText={tooltip}
          mode={mode}
          searchType={searchType}
        />
      </Grid>
      <Grid
        item
        xs={7}
        id={`${searchType}_search-input`}
        className={`${searchType}-search-input ${
          matchesQuery ? `${searchType}-search-input--mobile` : ''
        } ${styles.baseSearchInput}`}
      >
        <SearchInput
          mode={mode}
          setFocused={setFocused}
          view={view}
          inputPlaceholder={inputPlaceholder}
          searchType={searchType}
        />
        {searchResults.length > 0 && focused && (
          <SearchResultsList mode={mode} view={view} searchType={searchType} />
        )}
      </Grid>
    </Grid>
  );
}

export default SearchComponent;
