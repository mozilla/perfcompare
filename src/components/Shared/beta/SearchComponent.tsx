import {
  Dispatch,
  SetStateAction,
  useEffect,
  useState,
  createRef,
} from 'react';

import CheckIcon from '@mui/icons-material/Check';
import CloseIcon from '@mui/icons-material/Close';
import Button from '@mui/material/Button';
import Grid from '@mui/material/Grid';
import useMediaQuery from '@mui/material/useMediaQuery';

import { RootState } from '../../../common/store';
import { useAppDispatch } from '../../../hooks/app';
import { useAppSelector } from '../../../hooks/app';
import useSelectRevision from '../../../hooks/useSelectRevision';
import { clearCheckedRevisions } from '../../../reducers/CheckedRevisions';
import { SearchStyles } from '../../../styles';
import type { Revision } from '../../../types/state';
import type { Repository } from '../../../types/state';
import SearchDropdown from '../../Search/beta/SearchDropdown';
import SearchInput from '../../Search/beta/SearchInput';
import SearchResultsList from '../../Search/beta/SearchResultsList';

interface BaseSearchProps {
  mode: 'light' | 'dark';
  view: 'compare-results' | 'search';
  setPopoverIsOpen?: Dispatch<SetStateAction<boolean>>;
  prevRevision?: Revision;
  selectLabel: string;
  tooltip: string;
  inputPlaceholder: string;
  base: 'base' | 'new';
  repository: Repository['name'];
}

function SearchComponent({
  mode,
  view,
  setPopoverIsOpen,
  prevRevision,
  selectLabel,
  tooltip,
  base,
  inputPlaceholder,
  repository,
}: BaseSearchProps) {
  const [focused, setFocused] = useState(false);
  const dispatch = useAppDispatch();
  const { replaceSelectedRevision } = useSelectRevision();
  const matchesQuery = useMediaQuery('(max-width:768px)');
  const containerRef = createRef<HTMLDivElement>();
  const searchResults = useAppSelector(
    (state: RootState) => state.search.searchResults,
  );
  const baseSearchResults = useAppSelector(
    (state: RootState) => state.search.baseSearchResults,
  );
  const newSearchResults = useAppSelector(
    (state: RootState) => state.search.newSearchResults,
  );
  const searchState = useAppSelector((state: RootState) => state.search);
  const inputErrorBase = searchState.inputErrorBase;
  const inputErrorNew = searchState.inputErrorNew;
  const inputHelperText = searchState.inputHelperText;
  const styles = SearchStyles(mode);

  const handleFocus = (e: MouseEvent) => {
    if (
      (e.target as HTMLElement).matches(`#${base}-search-container, 
      #${base}-search-container *`) &&
      // do not open search results when dropdown or cancel button is clicked
      !(e.target as HTMLElement).matches(
        `#${base}_search-dropdown,
        #${base}_search-dropdown *,
        #cancel-edit-${base}-button, 
        #cancel-edit-${base}-button *`,
      )
    ) {
      setFocused(true);
      return;
    } else {
      setFocused(false);
    }
  };

  const handleEscKeypress = (e: KeyboardEvent) => {
    if (e.key === 'Escape') {
      setFocused(false);
    }
  };

  useEffect(() => {
    if (focused == false) {
      dispatch(clearCheckedRevisions());
    }
  }, [focused]);

  useEffect(() => {
    document.addEventListener('mousedown', handleFocus);
    return () => {
      document.removeEventListener('mousedown', handleFocus);
    };
  }, []);

  useEffect(() => {
    document.addEventListener('keydown', handleEscKeypress);
    return () => {
      document.removeEventListener('keydown', handleEscKeypress);
    };
  });

  return (
    <Grid
      container
      alignItems='flex-start'
      id={`${base}-search-container`}
      className={styles.container}
      ref={containerRef}
    >
      <Grid
        item
        xs={2}
        id={`${base}_search-dropdown`}
        className={`${base}-search-dropdown ${styles.dropDown} ${styles.baseSearchDropdown}`}
      >
        <SearchDropdown
          view={view}
          selectLabel={selectLabel}
          tooltipText={tooltip}
          mode={mode}
          base={base}
          repository={repository}
        />
      </Grid>
      <Grid
        item
        xs={7}
        id={`${base}_search-input`}
        className={`${base}-search-input ${
          matchesQuery ? `${base}-search-input--mobile` : ''
        } ${styles.baseSearchInput}`}
      >
        <SearchInput
          mode={mode}
          setFocused={setFocused}
          view={view}
          inputPlaceholder={inputPlaceholder}
          base={base}
          inputError={base === 'base' ? inputErrorBase : inputErrorNew}
          inputHelperText={inputHelperText}
        />
        {searchResults.length > 0 && focused && (
          <SearchResultsList
            mode={mode}
            searchResults={
              base === 'base' ? baseSearchResults : newSearchResults
            }
            view={view}
            base={base}
          />
        )}
      </Grid>

      <Grid item xs={7} className='compare-result-view'>
        {view == 'compare-results' && setPopoverIsOpen && prevRevision && (
          <>
            <Button
              className='edit-base-button'
              id={`replace-${base}-button`}
              data-testid={`replace-${base}-button`}
              size='small'
              onClick={() => replaceSelectedRevision(prevRevision)}
            >
              <CheckIcon className='accept' />
            </Button>
            <Button
              className={`edit-${base}-button`}
              id={`cancel-edit-${base}-button`}
              data-testid={`cancel-edit-${base}-button`}
              size='small'
              onClick={() => setPopoverIsOpen(false)}
            >
              <CloseIcon className='cancel' />
            </Button>
          </>
        )}
      </Grid>
    </Grid>
  );
}

export default SearchComponent;
