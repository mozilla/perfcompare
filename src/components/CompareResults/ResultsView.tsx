import { useEffect } from 'react';

import Grid from '@mui/material/Grid';
import { useLoaderData } from 'react-router-dom';
import { style } from 'typestyle';

import { useAppSelector } from '../../hooks/app';
import { SearchContainerStyles, background } from '../../styles';
import CompareWithBase from '../Search/CompareWithBase';
import { LinkToHome } from '../Shared/LinkToHome';
import PerfCompareHeader from '../Shared/PerfCompareHeader';
import type { LoaderReturnValue } from './loader';
import ResultsMain from './ResultsMain';

interface ResultsViewProps {
  title: string;
}
function ResultsView(props: ResultsViewProps) {
  const { baseRevInfo, newRevsInfo, frameworkId, results, baseRepo, newRepos } =
    useLoaderData() as LoaderReturnValue;

  const newRepo = newRepos[0];
  const { title } = props;
  const themeMode = useAppSelector((state) => state.theme.mode);
  const styles = {
    container: style({
      backgroundColor: background(themeMode),
    }),
  };

  const sectionStyles = SearchContainerStyles(themeMode, /* isHome */ false);

  useEffect(() => {
    document.title = title;
  }, [title]);

  return (
    <div
      className={styles.container}
      data-testid='beta-version-compare-results'
    >
      <PerfCompareHeader />
      <section className={sectionStyles.container}>
        <LinkToHome />

        <CompareWithBase
          hasNonEditableState={true}
          baseRev={baseRevInfo ?? null}
          newRevs={newRevsInfo ?? []}
          frameworkIdVal={frameworkId}
          isBaseSearch={null}
          expandBaseComponent={() => null}
          baseRepo={baseRepo}
          newRepo={newRepo}
        />
      </section>
      <Grid container alignItems='center' justifyContent='center'>
        <Grid item xs={12}>
          <ResultsMain results={results} />
        </Grid>
      </Grid>
    </div>
  );
}

export default ResultsView;
