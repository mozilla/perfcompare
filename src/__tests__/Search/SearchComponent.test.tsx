import { renderHook } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

import ResultsView from '../../components/CompareResults/ResultsView';
import SearchComponent from '../../components/Search/SearchComponent';
import { setSelectedRevisions } from '../../reducers/SelectedRevisionsSlice';
import { Strings } from '../../resources/Strings';
import useProtocolTheme from '../../theme/protocolTheme';

import getTestData from '../utils/fixtures';
import { renderWithRouter, store } from '../utils/setupTests';
import { screen } from '../utils/test-utils';



const protocolTheme = renderHook(() => useProtocolTheme()).result.current
  .protocolTheme;
const toggleColorMode = renderHook(() => useProtocolTheme()).result.current
  .toggleColorMode;

describe('Search Component', () => {


  it('RESULTS: shows dropdown and input when edit button in clicked', async () => {
    renderWithRouter(
      <ResultsView
        toggleColorMode={toggleColorMode}
        protocolTheme={protocolTheme}
        title='Results'
      />,
    );

    const user = userEvent.setup({ delay: null });
    const { testData } = getTestData();
    const selectedRevs = testData.slice(0, 2);

    act(() => {
      store.dispatch(setSelectedRevisions({ selectedRevisions: selectedRevs }));
    });

    const baseDropdown = document.querySelector(
      '.compare-results-base-dropdown',
    );

    expect(baseDropdown).not.toBeInTheDocument();

    const editButton = document.querySelector(
      '.edit-button-base',
    ) as HTMLElement;

    await act(async () => {
      await user.click(editButton);
    });

    expect(document.body).toMatchSnapshot();
    expect(screen.getByTestId('dropdown-select-base')).toBeInTheDocument();
    expect(screen.getAllByRole('textbox')[0]).toBeInTheDocument();
    const hiddenEditButton = document.querySelector('.hidden') as HTMLElement;

    expect(hiddenEditButton).toBeInTheDocument();

    await act(async () => void jest.runOnlyPendingTimers());
  });

  it('RESULTS: clicking the cancel button hides input and dropdown', async () => {
    renderWithRouter(
      <ResultsView
        toggleColorMode={toggleColorMode}
        protocolTheme={protocolTheme}
        title='Results'
      />,
    );

    const user = userEvent.setup({ delay: null });
    const { testData } = getTestData();
    const selectedRevs = testData.slice(0, 2);

    act(() => {
      store.dispatch(setSelectedRevisions({ selectedRevisions: selectedRevs }));
    });

    const baseDropdown = document.querySelector(
      '.compare-results-base-dropdown',
    );
    expect(baseDropdown).not.toBeInTheDocument();
    const editButton = document.querySelector(
      '.edit-button-base',
    ) as HTMLElement;

    await act(async () => {
      await user.click(editButton);
    });

    const cancelButton = document.querySelector(
      '.cancel-button',
    ) as HTMLElement;

    expect(document.body).toMatchSnapshot();
    expect(screen.getByTestId('dropdown-select-base')).toBeInTheDocument();
    expect(screen.getAllByRole('textbox')[0]).toBeInTheDocument();

    await act(async () => {
      await user.click(cancelButton);
    });

    expect(
      screen.queryByTestId('dropdown-select-base'),
    ).not.toBeInTheDocument();

    await act(async () => void jest.runOnlyPendingTimers());
  });

  it('RESULTS: clicking the save button hides input and dropdown', async () => {
    renderWithRouter(
      <ResultsView
        toggleColorMode={toggleColorMode}
        protocolTheme={protocolTheme}
        title='Results'
      />,
    );

    const user = userEvent.setup({ delay: null });
    const { testData } = getTestData();
    const selectedRevs = testData.slice(0, 2);

    await act(async () => {
      store.dispatch(setSelectedRevisions({ selectedRevisions: selectedRevs }));
    });

    const baseDropdown = document.querySelector(
      '.compare-results-base-dropdown',
    );
    const editButtonBase = document.querySelector(
      '.edit-button-base',
    ) as HTMLElement;
    expect(baseDropdown).not.toBeInTheDocument();

    await act(async () => {
      await user.click(editButtonBase);
    });

    const saveButtonBase = document.querySelector(
      '.save-button-base',
    ) as HTMLElement;

    expect(document.body).toMatchSnapshot();
    expect(screen.getByTestId('dropdown-select-base')).toBeInTheDocument();
    expect(screen.getAllByRole('textbox')[0]).toBeInTheDocument();

    await act(async () => {
      await user.click(saveButtonBase);
    });

    expect(
      screen.queryByTestId('dropdown-select-base'),
    ).not.toBeInTheDocument();

    //CHECK NEW SAVE BUTTON ACTION
    const editButtonNew = document.querySelector(
      '.edit-button-new',
    ) as HTMLElement;

    await act(async () => {
      await user.click(editButtonNew);
    });

    expect(screen.getByTestId('dropdown-select-new')).toBeInTheDocument();
    expect(screen.getAllByRole('textbox')[0]).toBeInTheDocument();

    const saveButtonNew = document.querySelector(
      '.save-button-new',
    ) as HTMLElement;

    await act(async () => {
      await user.click(saveButtonNew);
    });

    expect(screen.queryByTestId('dropdown-select-new')).not.toBeInTheDocument();
    await act(async () => void jest.runOnlyPendingTimers());
  });
});
