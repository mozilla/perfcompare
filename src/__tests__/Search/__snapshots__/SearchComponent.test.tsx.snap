// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search View renders correctly when there are no results 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root component_f14z0jsa css-plh9qo-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container container_f790aiy css-1ljzhs0-MuiGrid-root"
        id="base-search-container"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 base-search-dropdown dropDown_ffz7cf5 css-1awobkc-MuiGrid-root"
          id="base_search-dropdown"
        >
          <div>
            <div
              class="MuiFormControl-root search-dropdown fzaj1pt css-1nrlq1o-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined dropdown-select-label css-11mbixr-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                id="select-repository-label"
              >
                Base
                <svg
                  aria-hidden="true"
                  aria-label="The baseline revision (no changes) your revision will be compared against."
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall dropdown-info-icon css-1mecf1h-MuiSvgIcon-root"
                  data-mui-internal-clone-element="true"
                  data-testid="InfoOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                  />
                </svg>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl dropdown-select css-1q6aue5-MuiInputBase-root-MuiInput-root-MuiSelect-root"
                data-testid="dropdown-select-base"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="select-repository-label"
                  class="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input css-4yslue-MuiSelect-select-MuiInputBase-input-MuiInput-input"
                  role="button"
                  tabindex="0"
                >
                  try
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  tabindex="-1"
                  value="try"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-o3okz2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 base-search-input  baseSearchInput_fim9qh9 css-17gmabg-MuiGrid-root"
          id="base_search-input"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth fv98i7o css-q8hpuo-MuiFormControl-root"
          >
            <div
              class="hide"
            >
              Block
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root search-text-field base css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1vjgbo6-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-yzzuft-MuiInputAdornment-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
                    data-testid="SearchIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                    />
                  </svg>
                </div>
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1rgu65m-MuiInputBase-input-MuiOutlinedInput-input"
                  id="search-base-input"
                  name="baseSearch"
                  placeholder="Search base by ID number or author email"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      â€‹
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
